### ISSUE TITLE
Minimal Support for S3

### ISSUE THREAD

**Issue by renatocron** [pplu/aws-sdk-perl#63]
Hello,

How can I help you to add support to S3? 

Where you implement the code for 'fixing' things, like, transforming the HEAD request?

Paws::Net::APIRequest  {
    Parents       Moose::Object
    public methods (9) : content, header, header_hash, headers, meta, method, parameters, uri, url
    private methods (0)
    internals: {
        headers   :  HTTP::Headers,
        method    :  "HEAD",
        parameters:  {
            Bucket:  "foobucket",
            Key   :  "5a105e8b9d40e1329780d62ea2265d8a"
        },
        uri       :  "/foobucket/",
        url       :  "https://s3.amazonaws.com/foobucket/"
    }
}

   The url should have 5a105e8b9d40e1329780d62ea2265d8a in the end, for example.

Looking on Paws::SNS::Publish and Paws::SNS it looks that i'm missing something, because they work fine and I don't see any code for 'do request that way'

Thank you

**Comment by renatocron** [comment]
I updated my version and found S3APIRequest.pm.

I guess that where I need start

**Comment by pplu** [comment]
Hi Renato!

Sorry for the late response. I'm at the AWS global conference, so a bit busy... (will push a lot of new services soon!)

S3 is basically supported (I've had reports of people even succesfully using it for some choirs).

What's broken right now is GetObject and PutObject. When you upload something to S3 the signer doesn't seem to be doing a good job, so uploads fail, Operations that don't have a parameter sent in the body seem to work OK, so some diagnosing on what's happening at the signature phase would be great :)

Also, right now, "ParamInBody" parameters have to be strings, but these parameters represent big blobs, that we don't want necesarily in memory, so support for filehandle-like objects would be greatly appreciated.

Don't be afraid of touching whatever you want (even auto-generated classes), we'll work our way to get everything bak into master :smile:

**Comment by pplu** [comment]
Also: feel free to ask any question about code so I can guide you. Also take a look at this: http://www.slideshare.net/pplusdomain/paws-a-perl-aws-sdk. It may help you understand a bit more the internals.

**Comment by pplu** [comment]
Hi! Did you have any luck? Need any help?

**Comment by renatocron** [comment]
Hello Jose,

saw the slides, but had no chance to get your hands dirty yet.

I ended up using the Net::Amazon::S3 (old boy, ugly interface, but lot of
projects here using it) as the time and the task was critical.

I loved the simplicity of Paws, but at the moment, I can not help :/







Renato Santos
renato.santos@eokoe.com
(11) 3386-8181
www.eokoe.com
b-cam.com.br

**Comment by glenveegee** [comment]
Hi all, 

I've been looking into S3 support a bit because I'm rather desperate to get GetObject, PutObject and CopyObject working. If I want to submit a pull request is there a particular branch to submit it to or should I create a branch for the issue #63 (S3 related branch)

Regards, 

Glen

**Comment by pplu** [comment]
Try to take a hack at it in a branch :+1: the help you need along the way (point me to your fork if you need help). Take a look at the way the signing is done for S3 in the API specs, as it seems that S3 is a bit special.

Cheers!

**Comment by glenveegee** [comment]
Hi Jose I've made very small updates to restxml.pm in the builder, RestXmlCaller and S3APISignature and you can find my changes here https://github.com/glenveegee/aws-sdk-perl/tree/S3_support_63.

The changes correct a bug in the URL templating so that the correct request URL is sent to the API. It also adds  parameters that are defined in boto as part of a request and localized in the header to a request header. 

In an nutshell, there are a few fundamental deficiencies with the S3 request and signature. Most calls can work with anonymous requests but there are some (like copy operations - PUT and POST)  that have to be signed requests using the V4 algorithm and there appears to be no support for that in the S3 parts.

**Comment by pplu** [comment]
> In an nutshell, there are a few fundamental deficiencies with the S3 request and signature. Most calls can work with anonymous requests but there are some (like copy operations - PUT and POST) that have to be signed requests using the V4 algorithm and there appears to be no support for that in the S3 parts

If you want, change the signature role of the S3 service class to V4 to test out how that works. That way virtually every service will be on V4, and there is less code to maintain. The "special S3 signer" just comes from the fact that AWS specify a special signer in boto definition files.

**Comment by glenveegee** [comment]
Yes, I tried that but it's not part of the changes I pushed to my branch. I used Net::Amazon::Signature::V4 but no matter what I tried I couldn't create a valid signature. I'm not sure what I'm doing wrong on that front and it's not an easy problem to diagnose with the error amazon is sending back.

**Comment by pplu** [comment]
A couple of things that may help:

It looks like the keys have to be uriescaped between slashes

As in https://metacpan.org/source/Net::Amazon::S3::Request

Also you may find inspiration in how Net::Amazon::S3 signs requests

https://metacpan.org/source/RCONOVER/Net-Amazon-S3-0.80/lib/Net/Amazon/S3/HTTPRequest.pm

**Comment by glenveegee** [comment]
Hi Jose, 

I've finally managed to submit requests using the Net::Amazon::Signature::V4. Turns out it has a bug regarding how dates are handled and formatted. Anyway, my branch now restores the {Key+} requestURII syntax and handles it as in the boto python library. Also S3 API requests that use the restxml caller role now sign requests using the Net::Amazon::Signature::V4 (without modification). There are a few other changes here and there but FINALLY GetObject and CopyObject are working. I suspect this might fix many other methods but I'm not sure how to test this.

Let me know if you're interested in a pull request from  https://github.com/glenveegee/aws-sdk-perl/tree/S3_support_63.


Regards, 

Glen

**Comment by pplu** [comment]
Yay! :tada: :smile: :balloon: Excellent work!.

That looks like a tough bug to find and squat!

I'm surely interested in a PR from your branch!

**Comment by pplu** [comment]
I've pushed 0.20 to CPAN with the changes for S3 up to the moment. Have fun!

I'm leaving this issue open because S3 still needs development.

@glenveegee have you had time to test more?
